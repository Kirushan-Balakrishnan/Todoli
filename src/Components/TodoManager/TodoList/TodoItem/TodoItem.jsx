import React, { useState, useRef } from "react";
import { ListGroupItem, Button } from "reactstrap";
import {  editTodo } from "../../../../Services/TodoServices";
import { useStateValue } from "../../../../StateProvider/StateProvider";
import FiberManualRecordIcon from "@material-ui/icons/FiberManualRecord";
import "./TodoItem.css";






const TodoItem = ({todo, handleDelete}) => {
  console.log("Pending component rendered")
 
  const { title, status } = todo;
  const [{ started }, dispatch] = useStateValue();
  const [start, setStart] = useState(false);
  const [seconds, setSeconds] = useState(0);
  const [time, setTime] = useState("");
  const timer = useRef(null);


  




  const handleStart = () => {
    dispatch({ type: "SET_STARTED", bool: true });
    setStart(true);
    timer.current = setInterval(() => {
      setSeconds((currentTime) => currentTime + 1);
    }, 1000);
  };

  const handleStop = () => {
    clearInterval(timer.current);
    timer.current = null;
    setStart(false);
    setSeconds(0);
    setTime(formatTime())
    console.log(time)
    editTodo({ title, status: "completed", duration: formatTime() })
      .then((todo) => {
        dispatch({ type: "EDIT_TODO", todo: todo });
        dispatch({ type: "SET_STARTED", bool: false });
      })
      .catch((err) => console.log(err));
  };


  const handleDone = () =>{
    editTodo({ title, status: "completed", duration: null })
    .then((todo) => {
      dispatch({ type: "EDIT_TODO", todo: todo });
    })
    .catch((err) => console.log(err));
  }

   

  const formatTime = () => {
    console.log("Format time called")
    const getSeconds = `0${seconds % 60}`.slice(-2);
    const minutes = `${Math.floor(seconds / 60)}`;
    const getMinutes = `0${minutes % 60}`.slice(-2);
    const getHours = `0${Math.floor(seconds / 3600)}`.slice(-2);

    return `${getHours} : ${getMinutes} : ${getSeconds}`;
  }

  return (
    <ListGroupItem >
      <div className="main_div">
        <div className="main_flex">
          <div className="content_with_button">
            <div id='title_pending_todo' className="title_with_duration" aria-labelledby='title_pending_todo'>
              <label htmlFor='title_pending_todo' aria-labelledby='title_pending_todo'>{title}</label>
            </div>
          </div>
          <div className="content_with_button">
            {status === "pending" ? (
              !start ? (
                <>
                <Button outline color="info" size="sm" onClick={handleDone}>Mark As Done</Button>
                  <Button
                    style={{ marginRight: 128, marginLeft: 10 }}
                    outline
                    color="success"
                    onClick={handleStart}
                    size="sm"
                    disabled={started}
                  >
                    Start
                  </Button>
                  
                </>
              ) : (
                <>
                  <Button
                    style={{ marginLeft: 20 }}
                    outline
                    color="danger"
                    onClick={handleStop}
                    size="sm"
                  >
                    Stop
                  </Button>
                  <h1 style={{ marginLeft: 20 }}>{formatTime()}</h1>
                  <FiberManualRecordIcon
                    style={{ color: "green", marginLeft: 20 }}
                  />
                  <p style={{ color: "#636363", fontSize: 12 }}> Active</p>
                </>
              )
            ) : (
              ""
            )}
            {!start && (
              <button
                type="button"
                className="btn-close"
                aria-label="Close"
                style={{ marginLeft: 20 }}
                onClick={()=>handleDelete(todo)}
              />
            )}
          </div>
        </div>
      </div>
    </ListGroupItem>
  );
};

export default TodoItem;
