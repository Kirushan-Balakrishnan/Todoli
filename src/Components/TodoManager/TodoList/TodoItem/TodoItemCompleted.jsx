import React from "react";
import { ListGroupItem } from "reactstrap";
// import { deleteTodo } from "../../../../Services/TodoServices";
// import { useStateValue } from "../../../../StateProvider/StateProvider";
import "./TodoItem.css";

const TodoItemCompleted = React.memo(({ todo, handleDelete }) =>  {
  console.log("Completed component rendered")
 
  const { title, duration } = todo;
  // const [, dispatch] = useStateValue();

  

  return (
    <ListGroupItem >
      <div className="main_div">
        <div className="main_flex">
          <div className="content_with_button" htmlFor='completed_todo' aria-labelledby='completed_todo'>
            <div className="title_with_duration">
              <label htmlFor='completed_todo' aria-labelledby='completed_todo'>{title}</label>
              {duration != null ? (
                <p style={{ color: "#636363", fontSize: 12 }}>
                  {duration + " Time elapsed"}
                </p>
              ) : (
                ""
              )}
            </div>
          </div>

          <button
            type="button"
            className="btn-close"
            aria-label="Close"
            style={{ marginLeft: 20 }}
            onClick={()=>handleDelete(todo)}
          />
        </div>
      </div>
    </ListGroupItem>
  )
}) 

export default TodoItemCompleted;
