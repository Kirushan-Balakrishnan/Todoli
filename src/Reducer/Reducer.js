export const InitialState = {
  basket: [],
  started: false
};

// export const getBasketTotal = (basket) =>
// basket?.reduce((amount, item)=> item.price + amount, 0);

const Reducer = (state, action) => {
  switch (action.type) {
    
    case "SET_BASKET":
      return {
        ...state,
        basket: action.list,
      };
    case "EMPTY_BASKET":
      return {
        ...state,
        basket: [],
      };
    case "ADD_TO_BASKET":
      return {
        ...state,
        basket: [...state.basket, action.item],
      };
    case "REMOVE_FROM_BASKET":
      let newBasket = [...state.basket];

      const index = state.basket.findIndex(
        (basketItem) => basketItem.title === action.title
      );
      if (index >= 0) {
        newBasket.splice(index, 1);
      }
      return { ...state, basket: newBasket };
    case "EDIT_TODO":
      let exsisting = [...state.basket];
      const filtered = exsisting.filter((n) => n.title !== action.todo.title);
      filtered.push(action.todo);
      return { ...state, basket: filtered };
    case "SET_STARTED":
      return {
        ...state,
        started: action.bool,
      };
    default:
      return state;
  }
};

export default Reducer;
