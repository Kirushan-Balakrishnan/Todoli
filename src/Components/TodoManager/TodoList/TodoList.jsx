import React from "react";
import { ListGroup, Alert } from "reactstrap";
import { useStateValue } from "../../../StateProvider/StateProvider";
import TodoItem from "./TodoItem/TodoItem";
import CheckCircleIcon from "@material-ui/icons/CheckCircle";
import HighlightOffIcon from "@material-ui/icons/HighlightOff";
import TodoItemCompleted from "./TodoItem/TodoItemCompleted";
import { deleteTodo } from "../../../Services/TodoServices";



const TodoList = React.memo(() => {
  console.log("component rendered")

  const [{ basket }, dispatch] = useStateValue();
  const completed = basket.filter((n) => n.status === "completed");
  const pending = basket.filter((n) => n.status === "pending");


  const handleDelete  = React.useCallback((todo) => {
    deleteTodo(todo)
      .then((todo) => {
        dispatch({ type: "REMOVE_FROM_BASKET", title: todo.title });
      })
      .catch((err) => {
        console.log(err);
      });
  },[dispatch]) 


  if (basket.length === 0) {
    return <Alert>No Todos to show.</Alert>;
  }
  return (
    <>
      <label style={{ margin: 10 }}>Pending</label>

      <HighlightOffIcon style={{ color: "#d13030" }}></HighlightOffIcon>

      <ListGroup>
        {pending && pending.length === 0 ? (
          <p>No pending Todos to show</p>
        ) : (
          pending.map((todo) => <TodoItem key={todo.title} todo={todo} handleDelete={handleDelete} />)
        )}
      </ListGroup>

      <hr></hr>

      <label style={{ margin: 10 }}>Completed</label>

      <CheckCircleIcon style={{ color: "green" }}></CheckCircleIcon>

      <ListGroup>
        {completed && completed.length === 0 ? (
          <p>No completed Todos to show</p>
        ) : (
          completed.map((todo) => (
            <TodoItemCompleted key={todo.title} todo={todo}  handleDelete={handleDelete}/>
          ))
        )}
      </ListGroup>

    </>
  );
})

export default TodoList;
