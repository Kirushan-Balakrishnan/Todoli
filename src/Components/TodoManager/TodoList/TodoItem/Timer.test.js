import React from "react";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { StateProvider } from "../../../../StateProvider/StateProvider";
import Reducer from "../../../../Reducer/Reducer";
import TodoList from "../../TodoList/TodoList";
import TodoItem from "./TodoItem";
import { mount, shallow } from "enzyme";
import { configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
configure({ adapter: new Adapter() });

describe("Timer", () => {
  const Setup = () => {
    const InitialState = {
      basket: [
        { title: "Meeting with Anurag", status: "pending", duration: null },
      ],
      started: false,
    };
    render(
      <StateProvider initialState={InitialState} reducer={Reducer}>
        <TodoList />
      </StateProvider>
    );
  };

  beforeEach(() => {
    jest.useFakeTimers();
  });

  test("Start timer and stop timer", async () => {
    Setup();

    await act(async () => {
      userEvent.click(screen.getByRole("button", { name: /Start/i }));
    });

    expect(screen.getByRole("listitem")).toHaveTextContent("00 : 00 : 00");

    jest.advanceTimersByTime(1000);

    expect(screen.getByRole("listitem")).toHaveTextContent("00 : 00 : 01");

    jest.advanceTimersByTime(1000);

    expect(screen.getByRole("listitem")).toHaveTextContent("00 : 00 : 02");
    screen.debug();
  });


  
  
});
