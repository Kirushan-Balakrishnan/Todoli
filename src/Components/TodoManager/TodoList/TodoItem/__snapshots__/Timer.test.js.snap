// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer should match the snapshot 1`] = `"<li class=\\"list-group-item\\"><div class=\\"main_div\\"><div class=\\"main_flex\\"><div class=\\"content_with_button\\"><div id=\\"title_pending_todo\\" class=\\"title_with_duration\\" aria-labelledby=\\"title_pending_todo\\"><label for=\\"title_pending_todo\\" aria-labelledby=\\"title_pending_todo\\">Meeting with Anurag</label></div></div><div class=\\"content_with_button\\"><button type=\\"button\\" class=\\"btn btn-outline-info btn-sm\\">Mark As Done</button><button type=\\"button\\" style=\\"margin-right:128px;margin-left:10px\\" class=\\"btn btn-outline-success btn-sm\\">Start</button><button type=\\"button\\" class=\\"btn-close\\" aria-label=\\"Close\\" style=\\"margin-left:20px\\"></button></div></div></div></li>"`;
